### PR TITLE
read_sql/to_sql can accept database URI as con parameter

### LINKED ISSUE #10654: read_sql/to_sql should accept database URI as con parameter
Hello,

it will be nice if `read_sql` (and so `read_sql_query` and `read_sql_table` could directly accept database URI as `con` parameter
http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_sql.html

So we could do

    import pandas as pd
    uri = 'dialect+driver://username:password@host:port/database'
    query = 'SELECT * FROM table'
    df = pd.read_sql(query, uri)

instead of :

    import pandas as pd
    from sqlalchemy import create_engine
    uri = 'dialect+driver://username:password@host:port/database'
    engine = create_engine(uri)
    query = 'SELECT * FROM table'
    df = pd.read_sql(query, engine)

Kind regards

PS: this kind of idea is also used by Blaze / odo
http://blaze.pydata.org/
http://odo.readthedocs.org/

PS2: A similar approach could maybe also apply to `to_sql(...)`

### PR DESCRIPTION
**Pull request by s-celles**
It should fix https://github.com/pydata/pandas/issues/10654

### CODE REVIEW

**Comment by scls19fr** [comment]
I should have remove

    engine = sqlalchemy.create_engine(db_uri)

in `test_sql.py` `test_to_sql_read_sql_with_database_uri`

File: pandas/io/tests/test_sql.py [hunk 1]
```diff
@@ -918,6 +918,25 @@ def test_sqlalchemy_type_mapping(self):
         table = sql.SQLTable("test_type", db, frame=df)
         self.assertTrue(isinstance(table.table.c['time'].type, sqltypes.DateTime))
 
+    def test_to_sql_read_sql_with_database_uri(self):
+
+        # Test read_sql and .to_sql method with a database URI (GH10654)
+        from pandas import read_csv
+        iris_csv_file = os.path.join(tm.get_data_path(), 'iris.csv')
+        iris_frame1 = read_csv(iris_csv_file)
```
Comment (jorisvandenbossche) [inline pandas/io/tests/test_sql.py]: I think this is not really needed, you can just make a small dummy frame to test yourself, or maybe better use `self.test_frame1`

**Comment by jorisvandenbossche** [comment]
Looking good!

* We will have to remove the file 'iris.db' again afterwards. You can have a look at `tm.ensure_clean()` to use this.
* the `import sqlalchemy` will have to be wrapped in a try except block, as it is not guaranteed that this is installed. You have a `_SQLALCHEMY_INSTALLED` constant, but the problem is that this is only initialized the first time `_is_sqlalchemy_connectable` is used, and the `engine_builder` is called before that in case of `read_sql_table`

File: pandas/io/sql.py [hunk 2]
```diff
@@ -584,6 +589,18 @@ def has_table(table_name, con, flavor='sqlite', schema=None):
                   "MySQL will be further supported with SQLAlchemy connectables.")
 
 
+def engine_builder(con):
```
Comment (jorisvandenbossche) [inline pandas/io/sql.py]: As this is an internal function, let's name it starting with an underscore

**Comment by jorisvandenbossche** [comment]
@scls19fr To change things, you can just add more commits in the same branch and push again, then this PR will be automatically updated.

**Comment by scls19fr** [comment]
I think that's ok now but I may be wrong.

**Comment by TomAugspurger** [comment]
And we'd be grateful if you could squash these down to one commit. Let us know if you need any help: https://github.com/pydata/pandas/wiki/Using-Git#fetch-and-then-rebase-interactively-to-squash-reword-fix-otherwise-change-some-commits

**Comment by scls19fr** [comment]
I'm facing some issues to squash. Any help is welcome.

**Comment by jorisvandenbossche** [comment]
Normally, the following should do it:

```
git fetch upstream
git rebase -i upstream/master
```

Then you will see a pop-up window (or a vim session), where you can change occurences of 'pick' to 'squash' or 'fixup' 
And force push the branch again: `git push -f origin issue_10654`

But no worries, if you can't get it working, I can do it for you before merging

**Comment by scls19fr** [comment]
That's nice I'm with Tom on Gitter

**Comment by TomAugspurger** [comment]
@jorisvandenbossche I think we're good on the rebase.

**Comment by jorisvandenbossche** [comment]
@scls19fr Thanks a lot for your contribution!

**Pull request merged**